File: Algo/Leetcode/079PathInMatrix.cpp
class Solution {
public:
    bool exist(vector<vector<char>>& board, string word) {
        for(int i=0; i<board.size(); i++){
            for(int j=0; j<board[0].size(); j++){
                if(dfs(board, i, j, word, 0))
                    return true;
            }
        }
        return false;
    }

    bool dfs(vector<vector<char>>& board, int row, int col, string word, int start_idx){
        if (start_idx >= word.length())
            return true;
        if(row<0 || row>=board.size() || col<0 || col>=board[0].size() || word[start_idx]!=board[row][col])
            return false;

        char tmp = board[row][col];
        board[row][col] = '/';
        bool found = dfs(board, row-1, col, word, start_idx+1) ||
                     dfs(board, row+1, col, word, start_idx+1) ||
                     dfs(board, row, col-1, word, start_idx+1) ||
                     dfs(board, row, col+1, word, start_idx+1);
        board[row][col] = tmp;
        return found;
    }
};

File: Algo/Leetcode/079PathInMatrix.py
class Solution(object):
    def exist(self, board, word):
        """
        :type board: List[List[str]]
        :type word: str
        :rtype: bool
        """
        if len(word) == 0:
            return True

        n_row = len(board)
        n_col = len(board[0])
        for i in range(n_row):
            for j in range(n_col):
                if self.dfs(i, j, board, word):
                    return True

        return False

    def dfs(self, i, j, board, word):
        if len(word) == 0:
            return True

        n_row = len(board)
        n_col = len(board[0])
        if i >= n_row or i < 0 or j < 0 or j >= n_col:
            return False

        if board[i][j] == word[0]:
            tmp = word[0]
            board[i][j] = '#'
            find = self.dfs(i + 1, j, board, word[1:])
            if not find:
                find = self.dfs(i - 1, j, board, word[1:])
            if not find:
                find = self.dfs(i, j + 1, board, word[1:])
            if not find:
                find = self.dfs(i, j - 1, board, word[1:])
            board[i][j] = tmp
            return find
        else:
            return False

